bufferevent_ssl.c

```diff
@@ -268,31 +269,38 @@ do_read(struct bufferevent_ssl *bev_ssl, int n_to_read) {
 	if (n < 0)
 		return OP_ERR;
 
-	for (i=0; i<n; ++i) {
+	for (i = 0; i < n;) {
 		if (bev_ssl->bev.read_suspended)
 			break;
 		bev_ssl->ssl_ops->clear_error();
-		r = bev_ssl->ssl_ops->read(bev_ssl->ssl, space[i].iov_base, space[i].iov_len);
-		if (r>0) {
+		r = bev_ssl->ssl_ops->read(
+			bev_ssl->ssl, space[i].iov_base + len, space[i].iov_len - len);
+		if (r > 0) {
 			result |= OP_MADE_PROGRESS;
 			if (bev_ssl->read_blocked_on_write)
 				if (clear_rbow(bev_ssl) < 0)
 					return OP_ERR | result;
-			++n_used;
-			space[i].iov_len = r;
 			bev_ssl->ssl_ops->decrement_buckets(bev_ssl);
+			len += r;
+			if (space[i].iov_len - len > 0) {
+				continue;
+			} else {
+				space[i].iov_len = len;
+				len = 0;
+				++i;
+			}
 		} else {
 			int err = bev_ssl->ssl_ops->get_error(bev_ssl->ssl, r);
 			bev_ssl->ssl_ops->print_err(err);
-			if (bev_ssl->ssl_ops->err_is_ok(err) && result & OP_MADE_PROGRESS) {
+			if (result & OP_MADE_PROGRESS) {
```

Comment:
Yeah, I also thought about this (also if this will be a final patch then we can remove err_is_ok callback), but I'm not sure, is reading more data then SSL socket has should give an error (SSL_ERROR_SSL) ? If so, why? I'm seeing similar code in nginx for instannce.

Reply:
Honestly, I don't know what SSL error return in this case...
But if every test can run pass then I guess it should be like that. :smile: 
Are regress tests triggered manually? If so, can you help trigger them to see result?

Reply:
>Honestly, I don't know what SSL error return in this case...

What bothers me is this paragraph from `SSL_read(3)`:

```
       <= 0
           The read operation was not successful, because either the connection was closed, an error occurred or action must be taken by the calling process.  Call SSL_get_error(3) with the return value ret to find out the reason.

           Old documentation indicated a difference between 0 and -1, and that -1 was retryable.  You should instead call SSL_get_error() to find out if it's retryable.
```

> But if every test can run pass then I guess it should be like that

Well, we don't have that match tests, so I wouldn't say that :)

> Are regress tests triggered manually? If so, can you help trigger them to see result?

We have CI, but it is slow (due to public workers), so I did not trigger them yet.

To run tests locally you can do one of the following:
```
make -j100 verify
bin/regress # if cmake
test/regress # if autotools
```

Reply:
Btw, all tests are pass... what should we do next to be confident with PR?

Reply:
This looks OK, but I need to check the SSL code first (a) make sure that this is OK to use it this way, and b) why it shows errors, I guess it is somehow related to truncate attacks)

Reply:
Ok, I've added some comments into the code for the future readers. Thank you for the fix!